### ISSUE TITLE
Debug protocol is not 100% JSON compliant

### ISSUE THREAD

**Issue by fatcerberus** [Microsoft/ChakraCore#5262]
Don't know whether this is known or not but I figured I'd bring it up in case this was a bug or oversight: I've found that when inspecting values using the debugger API (for example, a result of JsDiagEvaluate), for at least `NaN` and +/- `Infinity`, these values are provided directly as the "value" property of the serialized object.  NaN and Infinity are not encodable in JSON and will be coerced to `null` (losing information) if the object is actually converted to JSON.

**Comment by MSLaguana** [comment]
I think this was a choice, but it's a little unclear from looking at the code. There is a `JsrtDebuggerObjectProperty::GetJSONObject` method which ends up calling into `JsrtDebugUtils::AddPropertyType` and that directly adds numeric properties to an object, but it also acknowledges that NaN and Infinity and -0 need special handling, and so for those is also adds a `display` property as well which I believe should be a string.

However, the result of `GetJSONObject` isn't JSON, it's an object, so this isn't exactly wrong, and it makes it convenient to fetch the values if doing in-process debugging. If you do `JSON.Stringify` the object then those problem numeric fields would be converted to `null`, but the `display` field should preserve the correct value. I haven't verified that this is the case though.

**Comment by fatcerberus** [comment]
I agree it’s very convenient. :smiley:  I just thought it might be a bug as there seemed to be a general tendency not to try to encode values at all which aren’t JSON compatible, e.g. there is no “value” field for `undefined`.

The reason I brought this up is that I currently take advantage of this behavior so I want to make sure it’s by-design so that a later bugfix doesn’t cause mysterious breakages in my debugger.

Thanks! :+1:

**Comment by MSLaguana** [comment]
@liminzhu do you know what the intended spec/contract here is?

**Comment by fatcerberus** [comment]
On a related note: `null` is serialized as type: object (mirroring `typeof`) but also has a nonzero “handle” field, which tripped me up for a bit as SSj was displaying null variables as objects.  It took me a little while to figure out what caused that.

**Comment by liminzhu** [comment]
[ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf) (JSON spec) definitely says, 

> Numeric values that cannot be represented as sequences of digits (such as Infinity and NaN) are not permitted. 

but the behavior does look intentional. + @kfarnung who is a lot more familiar with that part of the codebase for comments.

**Comment by kfarnung** [comment]
I think that the original intention was that these values would also have a "display" property that would take precedence, but I'd have to do some digging into the old, now deleted, node-chakracore code.  I don't think the serialization behavior is intentional, but probably never caused an issue in the overall flow.

I don't intend to change this behavior, we don't really depend on the objects being serializable anymore and do post processing to fit the wire protocol (currently CrDP).

Let me do a little more digging into how node/Chrome handle these cases and get back to you.

**Comment by fatcerberus** [comment]
@kfarnung Any idea why `null` values get a valid “handle” property?  As mentioned above, this makes them indistinguishable from actual objects over the wire (because typeof null === object) unless you also inspect the “display” string, which feels less than ideal (this string is ostensibly for display only)

**Comment by kfarnung** [comment]
It looks like every object gets a `handle` regardless of type.  It also looks like node adds a subtype for `null` values:

```json
{
    "name": "nullVal",
    "value": {
        "type": "object",
        "subtype": "null",
        "value": null
    },
    "writable": true,
    "configurable": true,
    "enumerable": true,
    "isOwn": true
}
```

There's definitely room for improvement here, for ChakraCore the configurable and enumerable values aren't provided from the API, so I just always set them to true currently.

To answer the original question, the CrDP protocol at least provides a string property called `unserializableValue` for the cases called out here (excerpt from node again):

```json
{
    "name": "nanVal",
    "value": {
        "type": "number",
        "unserializableValue": "NaN",
        "description": "NaN"
    },
    "writable": true,
    "configurable": true,
    "enumerable": true,
    "isOwn": true
}
```
